### PR TITLE
Solution for nearby neighborhoods

### PR DESCRIPTION
**Pull request by meghakrishnamurthy**
PR to solve the nearby neighborhoods questions.

1. Make Location.distanceTo() compute the great-circle distance so that LocationTest passes.
- Done. Used the cos formula as showcased in the wiki - https://en.wikipedia.org/wiki/Great-circle_distance

2. Make Search.near() return the neighborhoods closest to a location so that SearchTest passes.
- Done. To achieve this I have created 2 caches:
  - DistanceCache that would cache the distance between 2 locations. It uses a smart DistanceCacheKey that will ensure that location A to location B is the same as Location B to Location A - hence avoiding duplicate computations
  - NearestNeighborhoodCache that will cache the distance to all neighborhoods in a LinkedHashMap (so that it is sorted by distance)

Lookups use the cache, and if the values are not found - it will compute the entries and memoize the results, hence helping future requests. Also due to the distance cache duplicate calculations are avoided.

3. Computing distances can add up! Cache the results of your distance computations. How many computations did you save? How much time did you save?
- In my solution, a request to search n nearest locations will run through all the neighborhoods. So the test resulted in 301 calculations. But any future requests will pull this data from the cache and will avoid all computations. To demonstrate this I had modified the test to do the search 10 times and the number of computations was still 301.

4. What's the time complexity of your solution? Can you beat O(n)? Our example includes a few hundred neighborhoods. Can your solution scale to support millions?
The time complexity of my solution is O(N*Log(N)) where N is the number of locations. The Java stream uses TimSort - https://en.wikipedia.org/wiki/Timsort - which in the best case scenario can perform at O(N) but the average continues to be O(N*Log(N)). Having said that, this hit is only once since all the results are cached and most requests will take O(K) where K is the number of closest neighborhoods that need to be returned.

There are a few more optimizations that can be done, which would help the implementation:
- [ ] Use Guava cache and set a max on the number of objects in the cache to support scalability when the number of locations goes up to millions. Also the caching strategy I would propose would be an LRU cache
- [ ] I would also consider a distributed cache, these latitudes and longitudes do not change and there will not be a requirement to run these computations against each instance that is loaded up. We can avoid this by using a distributed cache

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -11,7 +11,7 @@
     <dependency>
       <groupId>com.google.guava</groupId>
       <artifactId>guava</artifactId>
-      <version>23.0</version>
+      <version>15.0</version>
```
Comment (crazybob) [inline pom.xml:14]: Curious what happened here. Are you running an older version of Java?

File: src/main/java/present/Location.java [hunk 2]
```diff
@@ -38,8 +46,100 @@ public double longitude() {
...
+                    Math.cos(currentLatitude) * Math.cos(otherLatitude) * Math.cos(currentLongitude - otherLongitude)
+    );
+
+    return angleInRadians * ONE_RADIAN_IN_KM;
+  }
+
+  /**
+   * Private inner class for the cache key for DISTANCE_CACHE
+   *
+   * The motivation of this class is because distance between Point 1 to 2 is the same as Point 2 to 1. So this overrides equals to handle that use-case
```
Comment (crazybob) [inline src/main/java/present/Location.java:97]: Commutative. Nice!

File: src/main/java/present/Location.java [hunk 3]
```diff
@@ -38,8 +46,100 @@ public double longitude() {
...
+      }
+
+      DistanceCacheKey distanceCacheKey = (DistanceCacheKey) object;
+
+      return (current.equals(distanceCacheKey.current) && other.equals(distanceCacheKey.other)) ||
+              (current.equals(distanceCacheKey.other) && other.equals(distanceCacheKey.current));
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (crazybob) [inline src/main/java/present/Location.java:135]: The prime isn't necessary here. We could use:
```
public int hashCode() {
  return Objects.hashCode(current) + Objects.hashCode(other);
}
```
Even better, we could defend against nulls in the constructor and have:
```
public int hashCode() {
  return current.hashCode() + other.hashCode();
}
```

File: src/main/java/present/Search.java [hunk 4]
```diff
@@ -15,6 +23,78 @@
    * @return the {@code n} nearest neighborhoods, ordered nearest to farthest
    */
   public static List<Neighborhood> near(Location location, int n) {
-    throw new UnsupportedOperationException("Please implement!");
+    Map<Neighborhood, Double> sortedNeighborhoodMap = getSortedNeighborhoodMapFromCache(location);
+    List<Neighborhood> nearestNeighborhoodList = new ArrayList<>();
```
Comment (crazybob) [inline src/main/java/present/Search.java:27]: Next time, we could use:

Lists.newArrayList(Iterables.limit(sortedNeighborhoodMap.keySet(), n));

File: src/main/java/present/Search.java [hunk 5]
```diff
@@ -15,6 +23,78 @@
...
+  }
+
+  /**
+   *
+   * Get a map of all neighborhoods and the distance to them from current location
+   *
+   * @param location to consider when computing distances to all other locations
+   * @return map of neighborhoods and the distance to them
+   */
+  private static Map<Neighborhood, Double> getNeighborhoodMap(Location location) {
```
Comment (crazybob) [inline src/main/java/present/Search.java:84]: A more descriptive name to consider: getDistancesByNeighborhood()

**Review by crazybob — COMMENTED** [review]
Great start!

It's funny you mention Guava's Cache and TimSort. I created Guava's Cache (originally called MapMaker), and Josh Bloch wrote Java's TimSort implementation while on my team at Android. I reviewed the code. 😋

How could we beat O(n*log(n)) on the first invocation (i.e. w/o caching the aggregate result)? How could we beat O(n)? Don't worry about implementing it. I'm just curious how you would go about it.

File: pom.xml (hunk 1 above)
Comment (meghakrishnamurthy) [reply pom.xml:14]: IntelliJ found this version. Did not bother investigating further.

File: src/main/java/present/Location.java (hunk 3 above)
Comment (meghakrishnamurthy) [reply src/main/java/present/Location.java:135]: Thank you for that. Let me know if you want me to change it.

File: src/main/java/present/Search.java (hunk 4 above)
Comment (meghakrishnamurthy) [reply src/main/java/present/Search.java:27]: This is good to know.

File: src/main/java/present/Search.java (hunk 5 above)
Comment (meghakrishnamurthy) [reply src/main/java/present/Search.java:84]: Sure. Makes sense.

**Comment by meghakrishnamurthy** [comment]
That's amazing that you built the Guava Cache. Definitely very useful! :)

The one solution that I can think of is:
-I was working on the Localized Grid search in the other code. We could potentially use that here. So instead of computing distances for all neighborhoods and sorting them. We can build up a grid with the lat / long. Once we have this setup in the grid - we do a grid search, if we find the required number of results then great, if not then search surrounding grids in an outward fashion till we get the required number of closest Neighborhoods. This would help reduce N - i.e the number of Neighborhoods we consider in each pass.

I will update if I can think of other solutions.

**Pull request closed without merging**